### ISSUE TITLE
BCP assertion failure if a NOT NULL datetime column with a default column is not provided

### ISSUE THREAD

**Issue by jmuehlner** [FreeTDS/freetds#478]
I have a SQL Server table like the following:

```
CREATE TABLE [Test] (
    TestID [int],
    TestDate [datetime] NOT NULL DEFAULT GETDATE()
);
```

I want to insert values into it using BCP, but without providing a value for `TestDate`:
```
bcp_init(dbproc, "Test", NULL, NULL, DB_IN);
bcp_bind(dbproc, (BYTE* ) "1", 0, 0, (BYTE *) "", 0, SYBTEXT, 1);
bcp_sendrow(dbproc);
```

At this point, I get this assertion failure:
```data.c:953: tds_generic_put: Assertion `curcol->column_varint_size' failed.```

I can insert into the table without issue using a regular INSERT statement:
`INSERT INTO Test(TestID) values("1");`

**Comment by blieusong** [comment]
Not too sure, but at least you should fix the `bcp_bind`. You are trying to put a string in an `integer` field

Maybe

```c
int test_id = 1;

bcp_init(dbproc, "Test", NULL, NULL, DB_IN);
bcp_bind(dbproc, (BYTE *)&test_id, 0, -1, NULL, 0, SYBINT4, 1);
bcp_sendrow(dbproc);
```

**Comment by jmuehlner** [comment]
Is binding a host string field to a remote numeric field not ok? The docs don't say anything about the types having to match, and it's always worked as expected when I've done it in the past.

In any case, I tested that change and it makes no difference to this problem. I'm still getting the `data.c:953: tds_generic_put: Assertion `curcol->column_varint_size' failed.` assertion.

**Comment by blieusong** [comment]
You are absolutely right... my (huge) mistake. I was just using that and have assumed the types should be identical.

I tried the first solution you mentioned with the official DB-library on a Sybase ASE server and it worked. The solution I propose obviously works too. I tried the same with freetds and it fails on `bcp_sendrow()`

**Comment by jmuehlner** [comment]
That makes me wonder which other column types are affected. I bet it's not just `datetime`.

**Comment by fziglio** [comment]
It would be nice to understand what Microsoft bcp utility will do. You are trying to insert a NULL value with a type that does not support NULL so the code is bailing out (not saying this is the best way to do it, at least error should be detected and reported).

**Comment by jmuehlner** [comment]
It looks to me like the microsoft `bcp` utility offers a way to do bulk imports where any unspecified values will use the default, which is what I would like to do as well.

https://learn.microsoft.com/en-us/sql/relational-databases/import-export/keep-nulls-or-use-default-values-during-bulk-import-sql-server?view=sql-server-ver16#using-bcp-and-using-default-values-without-a-format-file

**Comment by fziglio** [comment]
I was more curious on what at protocol level is happening. Does the protocol change or BCP is still used? It a NULL type used instead of a no-NULL type? For fixed value for Sybase the default value is extracted and used during BCP but in case of `DEFAULT GETDATE()` the default is generated by the server.

**Comment by blieusong** [comment]
The Sybase DB-lib documentation, which I use, specifies that setting `bcp_bind`'s variable length to zero means the value is NULL. This works fine with the official library but not with FreeTDS's `bcp_bind` (on an ASE 15.7 server). I haven't been able to check why.
